pkg/controllers/subnet/subnet_controller.go

```diff
@@ -50,102 +48,141 @@ type SubnetReconciler struct {
-	} else {
-		if controllerutil.ContainsFinalizer(obj, servicecommon.SubnetFinalizerName) {
-			metrics.CounterInc(r.SubnetService.NSXConfig, metrics.ControllerDeleteTotal, MetricResTypeSubnet)
-			if err := r.DeleteSubnet(*obj); err != nil {
-				log.Error(err, "deletion failed, would retry exponentially", "subnet", req.NamespacedName)
-				deleteFail(r, ctx, obj, err.Error())
-				return ResultRequeue, err
-			}
-			controllerutil.RemoveFinalizer(obj, servicecommon.SubnetFinalizerName)
-			if err := r.Client.Update(ctx, obj); err != nil {
-				log.Error(err, "deletion failed, would retry exponentially", "subnet", req.NamespacedName)
-				deleteFail(r, ctx, obj, err.Error())
-				return ResultRequeue, err
-			}
-			log.V(1).Info("removed finalizer", "subnet", req.NamespacedName)
-			deleteSuccess(r, ctx, obj)
-		} else {
-			log.Info("finalizers cannot be recognized", "subnet", req.NamespacedName)
+		log.Info("Updated Subnet CR", "Subnet", req.NamespacedName)
+	}
+
+	tags := r.SubnetService.GenerateSubnetNSTags(subnetCR)
+	if tags == nil {
+		log.Error(nil, "failed to generate Subnet tags", "Subnet", req.NamespacedName)
+		return ResultRequeue, errors.New("failed to generate Subnet tags")
+	}
+	// List VPC Info
+	vpcInfoList := r.VPCService.ListVPCInfo(req.Namespace)
+	if len(vpcInfoList) == 0 {
+		log.Info("No VPC info found, requeueing", "Namespace", req.Namespace)
+		return ResultRequeueAfter10sec, nil
+	}
+	// Create or update the subnet in NSX
+	if _, err := r.SubnetService.CreateOrUpdateSubnet(subnetCR, vpcInfoList[0], tags); err != nil {
+		if errors.As(err, &nsxutil.ExceedTagsError{}) {
+			log.Error(err, "tags limit exceeded, not retrying", "Subnet", req.NamespacedName)
+			updateFail(r, ctx, subnetCR, err.Error())
+			return ResultNormal, nil
 		}
+		log.Error(err, "failed to create/update Subnet, retrying", "Subnet", req.NamespacedName)
+		updateFail(r, ctx, subnetCR, err.Error())
+		return ResultRequeue, err
+	}
+	// Update status
+	if err := r.updateSubnetStatus(subnetCR); err != nil {
+		log.Error(err, "failed to update Subnet status, retrying", "Subnet", req.NamespacedName)
+		updateFail(r, ctx, subnetCR, err.Error())
+		return ResultRequeue, err
 	}
+	updateSuccess(r, ctx, subnetCR)
 	return ctrl.Result{}, nil
 }
 
-func (r *SubnetReconciler) DeleteSubnet(obj v1alpha1.Subnet) error {
-	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, string(obj.GetUID()))
-	if len(nsxSubnets) == 0 {
-		log.Info("no subnet found for subnet CR", "uid", string(obj.GetUID()))
-		return nil
-	}
-	portNums := len(r.SubnetPortService.GetPortsOfSubnet(*nsxSubnets[0].Id))
-	if portNums > 0 {
-		err := errors.New("subnet still attached by port")
-		log.Error(err, "", "ID", *nsxSubnets[0].Id)
+func (r *SubnetReconciler) deleteSubnetByID(subnetID string) error {
+	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, subnetID)
+	return r.deleteSubnets(nsxSubnets)
```

Comment:
Do we still need to list and check all the k8s crd ids when deleteSubnetByID?
Would you think it would be better to rename the current deleteSubnets to deleteStaleSubnets and have another function deleteSubnets only responsible for subnets deletion?